src/decode.jl

```diff
@@ -0,0 +1,147 @@
+# full decompression
+@btime jpeg_decode(filename); # 224.760 ms (7 allocations: 51.54 MiB)
+# only decompress luminance component
+@btime jpeg_decode(Gray, filename); # 91.157 ms (6 allocations: 17.18 MiB)
+# process only a few pixels
+@btime jpeg_decode(filename; scale_ratio=0.25); # 77.254 ms (8 allocations: 3.23 MiB)
+# process only a few pixels for luminance component
+@btime jpeg_decode(Gray, filename; scale_ratio=0.25); # 63.119 ms (6 allocations: 1.08 MiB)
+```
+
+"""
+function jpeg_decode(
+        ::Type{CT},
+        filename::AbstractString;
+        transpose=false,
+        scale_ratio=1) where CT<:Colorant
+    infile = ccall(:fopen, Libc.FILE, (Cstring, Cstring), filename, "rb")
+    @assert infile.ptr != C_NULL
+    out_CT, jpeg_cls = _jpeg_out_color_space(CT)
+
+    local cinfo, out
+    try
+        cinfo = LibJpeg.jpeg_decompress_struct()
+        cinfo_ptr = Ref(cinfo)
+        jerr = Ref{LibJpeg.jpeg_error_mgr}()
+        cinfo.err = LibJpeg.jpeg_std_error(jerr)
+        LibJpeg.jpeg_create_decompress(cinfo_ptr)
+        LibJpeg.jpeg_stdio_src(cinfo_ptr, infile)
+        LibJpeg.jpeg_read_header(cinfo_ptr, true)
+
+        # set decompression parameters, if given
+        r = _cal_scale_ratio(scale_ratio)
+        cinfo.scale_num, cinfo.scale_denom = r.num, r.den
+        cinfo.out_color_space = jpeg_cls
+
+        # eagerly calculate dimension information so that `output_XXX` fields are valid.
+        LibJpeg.jpeg_calc_output_dimensions(cinfo_ptr)
+        out_size = (Int(cinfo.output_width), Int(cinfo.output_height))
+        if !all(x -> x<=65535, out_size)
+            error("Suspicious inferred image size $out_size: each dimension is expected to have at most 65535 pixels.")
+        end
+        out_ndims = Int(cinfo.output_components)
+        @assert out_ndims == length(out_CT) "Suspicous output color space: $cinfo.out_color_space"
+
+        out = Matrix{out_CT}(undef, out_size)
+        _jpeg_decode!(out, cinfo)
+    finally
+        LibJpeg.jpeg_destroy_decompress(Ref(cinfo))
+        ccall(:fclose, Cint, (Ptr{Libc.FILE},), infile)
+    end
+
+    if out_CT <: CT
+        return transpose ? out : permutedims(out, (2, 1))
+    else
+        return transpose ? CT.(out) : CT.(PermutedDimsArray(out, (2, 1)))
+    end
+end
+function jpeg_decode(filename::AbstractString; kwargs...)
+    return jpeg_decode(_default_out_color_space(filename), filename; kwargs...)
+end
+
+function _jpeg_decode!(out::Matrix{<:Colorant}, cinfo::LibJpeg.jpeg_decompress_struct)
+    cinfo_ptr = Ref(cinfo)
+
+    row_stride = size(out, 1) * length(eltype(out))
+    buf = Vector{UInt8}(undef, row_stride)
+    out_uint8 = reinterpret(UInt8, out)
+
+    LibJpeg.jpeg_start_decompress(cinfo_ptr)
+    while cinfo.output_scanline < cinfo.output_height
+        # TODO(johnnychen94): try if we can directly write to `out` without using `buf`
+        GC.@preserve buf LibJpeg.jpeg_read_scanlines(cinfo_ptr, Ref(pointer(buf)), 1)
```

Comment:
```suggestion
        GC.@preserve buf LibJpeg.jpeg_read_scanlines(cinfo_ptr, pointer(buf), 1)
```

Reply:
I believe it should be `Ref(pointer(buf))` here:

```julia
const JSAMPLE = UInt8
const JSAMPROW = Ptr{JSAMPLE}
const JSAMPARRAY = Ptr{JSAMPROW}

function jpeg_read_scanlines(cinfo, scanlines, max_lines)
    ccall((:jpeg_read_scanlines, libjpeg), JDIMENSION, (Ptr{jpeg_decompress_struct}, JSAMPARRAY, JDIMENSION), cinfo, scanlines, max_lines)
end
```